### PR TITLE
Prompt beforeUnload (#4272)

### PR DESCRIPTION
**Pull request by timdorr**
Redo for #4359 with just the relevant commits.

@gpltaylor I'll give you commit access so you can provide any other cleanup or fixes that you want.

### CODE REVIEW

**Comment by timdorr** [comment]
Paging @pshrmn and @mjackson if they want to take a look now that the diff is cleaned up.

File: packages/react-router-dom/modules/BrowserPrompt.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    super(props)
+  }
+
+  onbeforeunload(e) {
+    var dialogText = "Changes you made may not be saved."
+    e.returnValue = dialogText
+    return dialogText
+  }
+
+  enableUnload(_onbeforeunload) {
```
Comment (pshrmn) [inline packages/react-router-dom/modules/BrowserPrompt.js]: Just a style critique, but I'm not crazy about underscore variable names.

File: packages/react-router-website/api/BrowserPrompt.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# BrowserPrompt
```
Comment (pshrmn) [inline packages/react-router-website/api/BrowserPrompt.md]: Missing `when` and `message` documentation.

File: packages/react-router-dom/modules/__tests__/BrowserPrompt-test.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import expect from 'expect'
+import React from 'react'
+import ReactDOM from 'react-dom'
+import { MemoryRouter } from '../../'
+import { Router } from '../../'
+import { Route } from '../../'
+import { BrowserPrompt } from '../../'
```
Comment (pshrmn) [inline packages/react-router-dom/modules/__tests__/BrowserPrompt-test.js]: These are importing from the build files, which I believe would require an `npm build` to be run before the tests are run. They should be importing from the source files, i.e., `import { MemoryRouter } from '../MemoryRouter'`

**Comment by gpltaylor** [comment]
@timdorr thanx for this.. been working on this all night. We never rebase at work so not use to doing this.

File: packages/react-router-dom/modules/BrowserPrompt.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
+import React, { PropTypes } from 'react'
+import { Prompt } from 'react-router'
+
+const BEFOREUNLOAD = "beforeunload"; // see: http://stackoverflow.com/questions/39094138/reactjs-event-listener-beforeunload-added-but-not-removed
```
Comment (pshrmn) [inline packages/react-router-dom/modules/BrowserPrompt.js]: I'm not sure what this is for. The stackoverflow question has to do with the function being the same, not the string `"beforeunload"`.

File: packages/react-router-dom/modules/BrowserPrompt.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+  static propTypes = {
+    when: PropTypes.bool,
+    message: PropTypes.oneOfType([
+        PropTypes.func,
+        PropTypes.string
+    ]).isRequired,
+    beforeUnload: PropTypes.oneOfType([
+        PropTypes.func,
+        PropTypes.bool
```
Comment (pshrmn) [inline packages/react-router-dom/modules/BrowserPrompt.js]: When you `beforeUnload` as a boolean value be used? Isn't that what the `when` prop is for?

File: packages/react-router-dom/modules/BrowserPrompt.js (hunk 4 above)
Comment (gpltaylor) [inline packages/react-router-dom/modules/BrowserPrompt.js]: I originally tried the window API "onbeforeunload" however, this did not work in React as we want unable to upload the event. Once the event was assigned, it was not possible to unregister.
https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onbeforeunload

It's force of habit to include reference to any solutions to issues we have for Sprint retros. This comment serves no purpose here. 

The comment is to show why I am using "old fashioned" string (something my team dislikes) events when there are dedicated "on" handlers built in.

File: packages/react-router-dom/modules/BrowserPrompt.js (hunk 5 above)
Comment (gpltaylor) [inline packages/react-router-dom/modules/BrowserPrompt.js]: I didn't want to lock the developer into an all or nothing.

For example, we might want to do the following.
- Stop navigating to different page and stop unload
- Allow navigation to different page but stop unload
- Stop navigation but conditionally allow unload 

I can see people using this as a general (all pages) onbeforeunload. If this is tied into the Prompt then prop, they would be unable to move to different pages.

File: packages/react-router-dom/modules/BrowserPrompt.js (hunk 1 above)
Comment (gpltaylor) [inline packages/react-router-dom/modules/BrowserPrompt.js]: Agreed, we used them to indicate a serious part of the (private) code that our juniors are not allowed to touch 😄

File: packages/react-router-website/api/BrowserPrompt.md (hunk 2 above)
Comment (gpltaylor) [inline packages/react-router-website/api/BrowserPrompt.md]: Yes, I didn't want to duplicate docs already within the Prompt. If we update Prompt we would have to remember to keep this up to date. That;s why I state that this is a wrapper to Prompt.

File: packages/react-router-dom/modules/__tests__/BrowserPrompt-test.js (hunk 3 above)
Comment (gpltaylor) [inline packages/react-router-dom/modules/__tests__/BrowserPrompt-test.js]: ok, I will fix

**Comment by gpltaylor** [comment]
hi @timdorr thank you for your patience. I am not familiar with the and procedures of Github, I really like React-Router and I want to be able to give a little back.

I have updated the code base and commented on some of the issues raised above.
How do I proceed? For example, how do we determine what work items (from the review) are to be done or ignored?

There's some conflict this morning, so I have rebased (successfully) and fixed. Have a good weekend.

**Comment by kdenz** [comment]
@gpltaylor Hi, does it work on iOS Safari? `Beforeunload` doesn't seem to work on iOS Safari, `pagehide` may be needed for it instead.

https://github.com/ReactTraining/react-router/issues/4267

File: packages/react-router-dom/modules/BrowserPrompt.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
+import React, { PropTypes } from 'react'
+import { Prompt } from 'react-router'
+
+const BEFOREUNLOAD = "beforeunload";
+
+var UnloadComponent = ComposedComponent => class extends React.Component {
```
Comment (mjackson) [inline packages/react-router-dom/modules/BrowserPrompt.js]: Can we please just use a regular component here? Higher-order components are unnecessary and just make stuff harder to read.

File: packages/react-router-dom/modules/BrowserPrompt.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    }).isRequired
+  }
+
+  static propTypes = {
+    when: PropTypes.bool,
+    message: PropTypes.oneOfType([
+        PropTypes.func,
+        PropTypes.string
+    ]).isRequired,
+    beforeUnload: PropTypes.oneOfType([
```
Comment (mjackson) [inline packages/react-router-dom/modules/BrowserPrompt.js]: Can we please call the prop `onBeforeUnload`?

File: packages/react-router-dom/modules/BrowserPrompt.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        PropTypes.func,
+        PropTypes.bool
+    ])
+  }
+
+  constructor(props) {
```
Comment (mjackson) [inline packages/react-router-dom/modules/BrowserPrompt.js]: No need for this constructor. Please remove it.

File: packages/react-router-website/examples/PreventingUnload.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
+import React from 'react'
+import Router from 'react-router-dom/BrowserRouter'
+import Route from 'react-router-dom/Route'
+import Link from 'react-router-dom/Link'
+import BrowserPrompt from '../../react-router-dom/BrowserPrompt'
```
Comment (mjackson) [inline packages/react-router-website/examples/PreventingUnload.js]: Please update the example to import everything from `react-router-dom` (see the other examples which have recently been updated.

File: packages/react-router-dom/modules/BrowserPrompt.js [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    super(props)
+  }
+
+  onbeforeunload(e) {
+    var dialogText = "Changes you made may not be saved."
```
Comment (mjackson) [inline packages/react-router-dom/modules/BrowserPrompt.js]: We use `const` everywhere. Please do the same here.

**Comment by gpltaylor** [comment]
Hi guys, been ill over the past week. Will look into this on the weekend.

**Comment by gpltaylor** [comment]
@timdorr @mjackson Code updated to match PR requirements Have a look and let me know what you think.

**Comment by gpltaylor** [comment]
@mjackson @timdorr changes requested pushed, can I get some feedback.
ta

**Comment by gpltaylor** [comment]
@mjackson @timdorr  can I get an update please?

**Comment by gpltaylor** [comment]
@mjackson poke

**Comment by timdorr** [comment]
Sorry, I don't have much of a dog in this fight, but it certainly *appears* to be OK to me.

**Comment by gpltaylor** [comment]
Thanx @timdorr

File: packages/react-router-dom/modules/BrowserPrompt.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    window.addEventListener(BEFOREUNLOAD, this.subscribed)
+  }
+
+  disableUnload() {
+    window.removeEventListener(BEFOREUNLOAD, this.subscribed)
+    delete this.subscribed
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.onBeforeUnload === true && this.props.onBeforeUnload === false) {
```
Comment (mjackson) [inline packages/react-router-dom/modules/BrowserPrompt.js]: We should be able to simplify the logic in this function.

File: packages/react-router-website/modules/components/APIDocs.js [hunk 12]
```diff
@@ -45,6 +45,14 @@ export const API = [
   { name: 'withRouter',
     html: require('../api/withRouter.md')
   },
+  { name: 'BrowserPrompt',
+    path: '/BroeserPrompt',
+    html: require('../api/BrowserPrompt.md')
+  },
+  { name: 'BrowserRouter',
```
Comment (mjackson) [inline packages/react-router-website/modules/components/APIDocs.js]: We don't need any changes in this file.

File: packages/react-router-website/modules/examples/PreventingUnload.js [hunk 13]
```diff
@@ -0,0 +1,87 @@
+import React from 'react'
+import {
+  BrowserRouter as Router,
+  Route,
+  Link,
+  Prompt,
+  BrowserPrompt
+} from 'react-router-dom'
+
+const PreventingTransitionsExample = () => (
```
Comment (gpltaylor) [inline packages/react-router-website/modules/examples/PreventingUnload.js]: If you click to navigate to google, it does it in the core browser, not the fake browser. So this event does fire. I would not release code without fully testing 😄 
Shall I take this out of the docs completely?

Comment (mjackson) [inline packages/react-router-website/modules/examples/PreventingUnload.js]: Ah, yes. I missed that ;)

But ya, I'd prefer to leave examples out of the "fake browser" that rely on stuff in the real browser.

**Comment by mjackson** [comment]
Also, I'm thinking this API should really just be a simple boolean value, instead of a function, e.g.

```js
<BrowserPrompt
  when={...}
  message={...}
  beforeUnload={true}
/>
```

The `beforeUnload` prop would just ensure we register for the `beforeunload` event. I don't think it needs to be a function.

**Comment by jochenberger** [comment]
I'm confused. Is it `onBeforeUnload` or `beforeUnload`? Also, how would I create a `Prompt` that only appears on unload but not on transition? `<Prompt when={false} (on)beforeUnload .../>`? This seems awkward to me.
Wouldn't it be more natural to me to have `when={({action})=> action === 'unload' }`?

**Comment by ryanflorence** [comment]
> Also, how would I create a Prompt that only appears on unload but not on transition?

You wouldn't use the router for that. Just register your own `window.onbeforeunload`.

**Comment by gpltaylor** [comment]
> Also, how would I create a Prompt that only appears on unload but not on transition?

> You wouldn't use the router for that. Just register your own window.onbeforeunload.

It is possible, you can conditionally stop (well alert) people when they try to move away from the site or refresh the page based on the prop passed to beforeUnload. 

`<Prompt beforeUnload={shouldBlock} />`

**Comment by gpltaylor** [comment]
@mjackson 
> Also, I'm thinking this API should really just be a simple boolean value, instead of a function, e.g.

This would make the implementation simple, however, it would stop people from being able to run "cleanup" code before they navigate away.

For example, I have systems that calls an APIs to save users details before they leave the page. Passing a function to this module allows it to be called.

Can you please confirm that you want this taking out. If so I would be happy to do so.
Also, after the changes, do you want me to squash the commits? There's a lot of nonsense in this branch now.

**Comment by mjackson** [comment]
> it would stop people from being able to run "cleanup" code before they navigate away.

The job of `<BrowserPrompt>` is just to prompt people before navigation happens. It shouldn't be considered a general-purpose hook for running code before the tab closes. If you want to do that, you should just register your own `window.onbeforeunload`.

> do you want me to squash the commits?

Yes, that would be great. Either way we'll probably squash everything down into a single commit when we merge.

File: packages/react-router-dom/modules/BrowserPrompt.js [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+        }).isRequired
+    }).isRequired
+  }
+
+  static propTypes = {
+    when: PropTypes.bool,
+    message: PropTypes.oneOfType([
+        PropTypes.func,
+        PropTypes.string
+    ]).isRequired,
```
Comment (ykzts) [inline packages/react-router-dom/modules/BrowserPrompt.js]: this prop has been used?

**Comment by bitcloud** [comment]
Perhaps I am missing something, but in the current incarnation this only triggers when you navigate outside the app e.g. without history API.
If you use the router Link Components and history.push the page is never unloaded so this event never triggers. This makes this component as is pretty useless or am I missing something?

**Comment by Kerumen** [comment]
@timdorr Are you still working on this?

**Comment by gpltaylor** [comment]
Hi @timdorr yes, I think I will update the changes requested an give this dead horse one more try :)
If not I will ping you to close. Thanx for your help.

**Comment by wldcordeiro** [comment]
@mjackson I just opened an issue because I was wanting this kind of behavior! 😄 If this PR has stalled out I could open a new one that incorporates the review and fixes the merge conflict.

**Comment by jimthedev** [comment]
Just a reminder that this is also a thing: https://github.com/xJkit/react-goodbye

**Comment by ibobo** [comment]
When will this be merged?

**Pull request closed without merging**